code/controllers/stonedmc/subsystem/codex.dm

```diff
@@ -0,0 +1,139 @@
+SUBSYSTEM_DEF(codex)
+	name = "Codex"
+	flags = SS_NO_FIRE
+	init_order = INIT_ORDER_CODEX
+	var/list/entries_by_path = list()
+	var/list/entries_by_string = list()
+	var/list/index_file = list()
+	var/list/search_cache = list()
+	var/list/entry_cache = list()
+
+/datum/controller/subsystem/codex/Initialize()
+
+	// Create general hardcoded entries.
+	for(var/ctype in typesof(/datum/codex_entry))
+		var/datum/codex_entry/centry = ctype
+		if(initial(centry.display_name) || initial(centry.associated_paths) || initial(centry.associated_strings))
+			centry = new centry()
+			for(var/associated_path in centry.associated_paths)
+				entries_by_path[associated_path] = centry
+			for(var/associated_string in centry.associated_strings)
+				entries_by_string[associated_string] = centry
+			if(centry.display_name)
+				entries_by_string[centry.display_name] = centry
+
+	// Create categorized entries.
+	for(var/ctype in subtypesof(/datum/codex_category))
+		var/datum/codex_category/cat = new ctype
+		cat.Initialize()
+		qdel(cat)
+
+	// Create the index file for later use.
+	for(var/thing in SScodex.entries_by_path)
+		var/datum/codex_entry/entry = SScodex.entries_by_path[thing]
+		index_file[entry.display_name] = entry
+	for(var/thing in SScodex.entries_by_string)
+		var/datum/codex_entry/entry = SScodex.entries_by_string[thing]
+		index_file[entry.display_name] = entry
+	index_file = sortTim(index_file, cmp=/proc/cmp_text_asc)
+	. = ..()
+
+/datum/controller/subsystem/codex/proc/get_codex_entry(var/datum/codex_entry/entry)
+	if(!initialized)
+		return
+	var/searching = "\ref[entry]"
+	if(!entry_cache[searching])
+		if(istype(entry))
+			entry_cache[searching] = entry
+		else
+			entry_cache[searching] = FALSE
+			if(ispath(entry))
+				entry_cache[searching] = entries_by_path[entry]
+			else if(istype(entry, /atom))
+				var/atom/entity = entry
+				if(entries_by_string[lowertext(entity.name)])
+					entry_cache[searching] = entries_by_string[lowertext(entity.name)]
+				else if(entries_by_path[entity.type])
+					entry_cache[searching] = entries_by_path[entity.type]
+				else if(entity.get_specific_codex_entry())
+					entry_cache[searching] = entity.get_specific_codex_entry()
+	return entry_cache[searching]
+
+/datum/controller/subsystem/codex/proc/present_codex_entry(var/mob/presenting_to, var/datum/codex_entry/entry)
+	if(entry && istype(presenting_to) && presenting_to.client)
+		var/list/dat = list()
+		if(entry.lore_text)
+			dat += "<font color='#abdb9b'>[entry.lore_text]</font>"
+		if(entry.mechanics_text)
+			dat += "<h3>OOC Information</h3>"
+			dat += "<font color='#9ebcd8'>[entry.mechanics_text]</font>"
+		if(entry.antag_text && presenting_to.mind && player_is_antag(presenting_to.mind))
```

Comment:
Be aware that I'm removing special_role and that the whole antag thing will most likely not work/conflict/break.